liquibase-standard/src/main/java/liquibase/snapshot/SnapshotGeneratorChain.java

```diff
@@ -59,39 +54,18 @@ public <T extends DatabaseObject> T snapshot(T example, DatabaseSnapshot snapsho
             return null;
         }
 
-        boolean firstRoundDone = false;
-        SnapshotGenerator lastGenerator = null;
-        T lastObject = example;
+        T objectToSnapshot = example;
         while (snapshotGenerators.hasNext()) {
             SnapshotGenerator generator = snapshotGenerators.next();
             if (replacedGenerators.contains(generator.getClass())) {
                 continue;
             }
-            T object = generator.snapshot(lastObject, snapshot, this);
+            T object = generator.snapshot(objectToSnapshot, snapshot, this);
             if ((object != null) && (object.getSnapshotId() == null)) {
                 object.setSnapshotId(snapshotIdService.generateId());
             }
-            // only first generator in the chain is allowed to create new instances - subsequent ones are not
-            if (firstRoundDone && object != lastObject) {
-                throw new DatabaseException(String.format("Snapshot generator %s has returned a different reference from the previous generator %s.\n" +
-                                "\tSnapshot object was: %s, it is now: %s.\n" +
-                                "\tConsider declaring %1$s as being replaced by one the generator in the chain via liquibase.snapshot.SnapshotGenerator#replaces.",
-                        generator.getClass().getName(),
-                        lastGenerator.getClass().getName(),
-                        identity(lastObject),
-                        identity(object)));
-            }
-            lastObject = object;
-            lastGenerator = generator;
-            firstRoundDone = true;
-        }
-        return lastObject;
-    }
-
-    private static String identity(Object object) {
-        if (object == null) {
-            return "null";
+            objectToSnapshot = object;
```

Comment:
this works only if all the generators carefully copy the existing attributes, right? (or are ordered properly)

Reply:
@filipelautert do you think it's worth documenting these "rules" somewhere?

Reply:
Yes, in the code and in the contribute site!